README.md
IF THIS FILE HAS NO LINE BREAKS:  View it in a web browser.  
Windows Notepad is a very popular text editor that comes with the 
Windows operating system, but it doesn't recognize line breaks in 
text files made on Linux or any other UNIX-like operating system.  
Text files with UNIX line breaks display correctly in WordPad, 
Notepad++, Programmer's Notepad, Gedit, or a web browser, just 
not in Windows Notepad.


This is a minimal working program for the Super Nintendo
Entertainment System.

Concepts illustrated:

* internal header and init code
* setting up a static background
* loading data from multiple 32 KiB banks in a LoROM
* structure of a game loop
* automatic controller reading
* 8.8 fixed-point arithmetic
* acceleration-based character movement physics
* sprite drawing and animation, with horizontal flipping
* makefile-controlled conversion of graphics to tile data in
  both 2-bit-per-pixel and 4-bit-per-pixel formats
* booting the SPC700 audio CPU
* writing SPC700 code using 65C02 syntax using blargg's
  [SPC700 macro pack] for ca65
* basic use of SPC700 timers to control playback
* makefile-controlled compression of sampled sound to BRR format
* creating an SPC700 state file for SPC music players

[SPC700 macro pack]: http://forums.nesdev.com/viewtopic.php?p=121690#p121690

Setting up the build environment
================================

Building this demo requires that the following software be
installed on your computer first:

* ca65 and ld65, the assembly language tools that ship
  with the cc65 C compiler
* Python, a programming language interpreter
* Python Imaging Library, a Python extension to read and write
  bitmap images
* GNU Make, a program to calculate which files need to be
  rebuilt when other files change
* GNU Coreutils, a set of simple command-line utilities for
  file management and text processing

It also requires general familiarity with the command prompt.
You are encouraged to read and understand the articles on general
computer science topics listed at [Before the basics].

[Before the basics]: http://wiki.nesdev.com/w/index.php/Before_the_basics

On Linux
--------

To install Make, Python 2, and Python Imaging Library under Ubuntu:

1. Open a terminal.
2. Type the following, followed by the Enter key:

        sudo apt-get install build-essential python-imaging

3. Type your password to authorize the installation.
4. The wave to BRR converter can optionally use the Numerical Python
   module to provide a slight speedup for the preemphasis step.
   Repeat steps 2 and 3, typing this:

        sudo apt-get install python-numpy

To install Make, Python 2, and Python Imaging Library under Fedora:
(instructions suggested by 43110; not tested)

1. Open a terminal and use `su` to become root.
2. Type the following, followed by the Enter key:

        yum install make automake gcc gcc-c++ python python-imaging

Because cc65 is a fairly niche tool, and because the C compiler
portion (which this demo does not use) is not free software as
defined by GNU and Debian, your Linux distribution's default
repository is unlikely to contain it.  This means you will
probably need to install it from source code.

1. Visit [cc65 on GitHub].
2. Click Download ZIP
3. Unzip into a new folder.
4. In a terminal issue the following commands (suggested by 43110):

        cd [path to where you unzipped cc65]
        make
        make install prefix=~/usr

   There's no `./configure` step, and `make install` reportedly
   complained about there being no prefix var. (case sensitive?)

5. Insert the following in your `.bash_profile` or `.bashrc` file,
   to automatically add the local executables to your PATH
   
        if [ -d "$HOME/usr/bin" ] ; then
            PATH="$HOME/usr/bin:$PATH"
        fi

[cc65 on GitHub]: https://github.com/cc65/cc65

On Windows
----------

The MSYS project ports Make, Coreutils, and other key parts of
the GNU operating environment to Windows.  An easy way to install
MSYS is through the automated installer provided by devkitPro.

1. Visit [devkitPro Getting Started].
2. Follow the instructions there to download and run the
   devkitPro Automated Installer.
3. In the installer, check only MSYS.  Don't check devkitARM,
   devkitPPC, devkitPSP, or any of the libraries for newer platforms
   unless you want to start developing for one of those soon.
   
To install Python under Windows:

1. Visit [Python home page].
2. Under Downloads, under Download for Windows, click Python 2.7.8.
   (This step may change if the maintainers of the Python web site
    reorganize the Python web site.)
3. In your web browser's downloads folder, run `python-2.7.8.msi`
4. Follow the prompts through the installer wizard.

To install Python Imaging Library under Windows:

1. Visit [Pillow project page].
2. Under Package, click the latest version.
3. Click Downloads.
4. Click `Pillow-2.5.3.win32-py2.7.exe`. The `py-2.7` part MUST
   match the major and minor version of Python installed on your PC.
   (In Python 2.7.8, the major version is 2, the minor version is 7,
   and the bug fix version is 8.)
5. In your web browser's downloads folder, run
   `Pillow-2.5.3.win32-py2.7.exe`
6. Follow the prompts through the installer wizard

To install cc65 under Windows:

1. Visit [cc65 introduction].
2. Scroll to the bottom.
3. Click "Windows Snapshot" to download a zip file.
4. Open the zip file.
5. Inside the zip file, open the bin folder.
6. Drag `ca65.exe` and `ld65.exe` into a new folder.
7. Add this folder to your "Path" environment variable.

Finally, to make ca65, ld65, and Python available to Make, you'll
need to add the folders containing `ca65.exe`, `ld65.exe`, and
`python.exe` to the `Path` environment variable.  Because the steps
for setting environment variables differ between versions of Windows,
you'll want to search the web for `windows x.x path variable`,
replacing `x.x` with `7`, `8.1`, etc.

Finally, open the makefile in a text editor and change EMU to
whatever Super NES emulator you have installed.  My Linux laptop
runs Martin Korth's NO$SNS emulator under Wine.  But if you have any
other emulator setup, you'll need to move the 2 lines that begin
with `run:` above the line that begins with `nocash-run:`.

[devkitPro Getting Started]: http://devkitpro.org/wiki/Getting_Started
[Python home page]: https://www.python.org/
[Pillow project page]: https://pypi.python.org/pypi/Pillow
[cc65 introduction]: http://cc65.github.io/cc65/

Organization of the program
===========================

Include files
-------------

* `snes.inc`: Register definitions and useful macros for the S-CPU
* `global.inc`: S-CPU global variable and function declarations
* `spc-ca65.inc`: Macro library to produce SPC700 instructions
* `spc-65c02.inc`: Macro library to use 65C02 syntax with the SPC700

Source code files (65816)
-------------------------

* `snesheader.s`: Super NES internal header
* `init.s`: PPU and CPU I/O initialization code
* `main.s`: Main program
* `bg.s`: Background graphics setup
* `player.s`: Player sprite graphics setup and movement
* `ppuclear.s`: Useful subroutines for interacting with the S-PPU
* `blarggapu.s`: Send a sound driver to the S-SMP 

Source code files (SPC700)
--------------------------

* `spcheader.s`: Header for the `.spc` file; unused in `.sfc`
* `spcimage.s`: Sound driver

Each source code file is made up of subroutines that start with
`.proc` and end with `.endproc`.  See the [ca65 Users Guide] for
what these mean.

[ca65 Users Guide]: http://cc65.github.io/doc/ca65.html

The tools
=========

The `tools` folder contains a few essential command-line programs
written in Python to convert asset data (graphics, audio, etc.) into
a form usable by the Super NES.  The makefile contains instructions
to run these programs whenever the original asset data changes.

* `pilbmp2nes.py` converts bitmap images in PNG or BMP format
  into tile data usable by several classic video game consoles.
  It has several options to control the data format; use
  `pilbmp2nes.py --help` from the command prompt to see them all.
* `wav2brr.py` converts an uncompressed wave file to the BRR (bit
  rate reduction) format, a lossy audio codec based on ADPCM used
  by the S-DSP (the audio chip in the Super NES).
* `mktables.py` generates some lookup tables, such as the
  pitch table to be used by a music engine.

Greets
======

* [Super Nintendo Development Wiki] contributors
* Martin Korth (nocash) for [Fullsnes] doc and [NO$SNS] emulator
* Shay Green (blargg) for APU examples
* Jeremy Chadwick (koitsu) for more code organization tips

[Super Nintendo Development Wiki]: http://wiki.superfamicom.org/
[Fullsnes]: http://problemkaputt.de/fullsnes.htm
[NO$SNS]: http://problemkaputt.de/sns.htm

Legal
=====

The demo is distributed under the following license, based on the
GNU All-Permissive License:

> Copyright 2014-2015 Damian Yerrick
> 
> Copying and distribution of this file, with or without
> modification, are permitted in any medium without royalty provided
> the copyright notice and this notice are preserved in all source
> code copies.  This file is offered as-is, without any warranty.



CHANGES.txt
0.05 (2015-02-13): The "Ho-ly-$#!+" Code Review
* fixed branch bug in collision introduced during translation
  from 6502 to "idiomatic" 65816 (reported by thefox)
* added overall map of source code to README.md
  (requested by Espozo)
* comment readability improvements
* added doc comments before more functions (requested by Espozo)
* more consistent use of seta/setxy macros
  to change processor word size
* split code related to the background and the player
  into separate files (requested by Espozo)
* compensated for missing background line 0
* explained unfamiliar register names
* renamed .h files to .inc so as not to confuse with C headers
  (requested by koitsu)
* rewrote init code based on koitsu's InitializeSNES macro
* indented forward branches (requested by Espozo)
* moved ROM segments to $808000 to allow use of fast ROM
  (requested by koitsu)
* imported the direct page base address as a symbol
  (requested by koitsu)
* broke snes.inc into functional groups
* added register name aliases from Martin Korth's fullsnes.htm
  (requested by koitsu)

0.04a (2015-01-28)
* corrected missing lorom256k.cfg file

0.04 (2014-11-09):
* expanded to 256 KiB to demonstrate far operation
* moved parts into separate banks to show use of multiple banks
* snesheader: added symbolic names for mappers, regions, and
  memory sizes and speeds
* wav2brr.py: added proper command line options including
  decompression of BRR back to WAV
* spcimage.s: fixed stya macro (reported by doppel)

0.03 (2014-10-03):
* wav2brr.py: made NumPy dependency optional
* added other explanatory comments to main.s
* Corrected some errors related to BSS in linker config file
* snes.h: macros for controlling accumulator and index width
  (requested by Bisqwit)
* snes.h: macro for nametable X, Y location (requested by Bisqwit)
* New template for SPC files

0.02 (2014-09-24): 
* converted README to Markdown
* added instructions for installing build environment to README
  (requested by whicker)
* corrected a few factual errors in README related to differences
  between the NES and the Super NES
* added a wav file and wav to BRR converter
* removed Encyclopedia Dramatica "goes where" meme from index.txt

0.01 (2014-09-21): 
* initial release


src/main.s
.include "snes.inc"
.include "global.inc"
.smart
.export main, nmi_handler

USE_PSEUDOHIRES = 0
USE_INTERLACE = 0

.segment "ZEROPAGE"
nmis: .res 1
oam_used: .res 2

.segment "BSS"

OAM:   .res 512
OAMHI: .res 512
; OAMHI contains bit 8 of X and the size bit for each sprite.
; It's a bit wasteful of memory, as the 512-byte OAMHI needs to be
; packed by software into 32 bytes before being sent to the PPU, but
; it makes sprite drawing code much simpler.  The OBC1 used in the
; game Metal Combat: Falcon's Revenge performs the same packing
; function in hardware, possibly as a copy protection method.

.segment "CODE"
;;
; Minimalist NMI handler that only acknowledges NMI and signals
; to the main thread that NMI has occurred.
.proc nmi_handler
  seta16
  phb
  phk         ; set data bank to bank 0 (because banks $40-$7D
  plb         ; and $C0-$FF can't reach low memory)
  inc a:nmis
  bit a:NMISTATUS
  plb
  rti
.endproc

;;
; This program doesn't use IRQs either.
.proc irq_handler
  rti
.endproc

.segment "CODE1"
; init.s sends us here
.proc main

  ; In the same way that the CPU of the Commodore 64 computer can
  ; interact with a floppy disk only through the CPU in the 1541 disk
  ; drive, the main CPU of the Super NES can interact with the audio
  ; hardware only through the sound CPU.  When the system turns on,
  ; the sound CPU is running the IPL (initial program load), which is
  ; designed to receive data from the main CPU through communication
  ; ports at $2140-$2143.  Load a program and start it running.
  jsl spc_boot_apu

  jsl load_bg_tiles  ; fill pattern table
  jsl draw_bg        ; fill nametable
  jsl load_player_tiles
  
  phk
  plb
  
  ; Program the PPU for the display mode
  seta8
  stz BGMODE     ; mode 0 (four 2-bit BGs) with 8x8 tiles
  stz BGCHRADDR  ; bg planes 0-1 CHR at $0000
  lda #$4000 >> 13
  sta OBSEL      ; sprite CHR at $4000, sprites are 8x8 and 16x16
  lda #>$6000
  sta NTADDR+0   ; plane 0 nametable at $6000
  sta NTADDR+1   ; plane 1 nametable also at $6000
  ; set up plane 0's scroll
  stz BGSCROLLX+0
  stz BGSCROLLX+0
  lda #$FF
  sta BGSCROLLY+0  ; The PPU displays lines 1-224, so set scroll to
  sta BGSCROLLY+0  ; $FF so that the first displayed line is line 0
  
.if ::USE_PSEUDOHIRES
  ; set up plane 1's scroll, offset by 4 pixels, to show
  ; the half-pixels of pseudohires
  lda #4
  sta BGSCROLLX+2
  stz BGSCROLLX+2
  lda #$FF
  sta BGSCROLLY+2
  sta BGSCROLLY+2
  lda #%00000010   ; enable plane 1 for left halves
  sta BLENDSUB
phbit = SUB_HIRES  ; split horizontal pixels
.else
phbit = 0
.endif

.if ::USE_INTERLACE
ilbit = INTERLACE
.else
ilbit = 0
.endif

  lda #phbit|ilbit
  sta PPURES
  lda #%00010001  ; enable sprites and plane 0
  sta BLENDMAIN
  lda #VBLANK_NMI|AUTOREAD  ; but disable htime/vtime IRQ
  sta PPUNMI

  ; Set up game variables, as if it were the start of a new level.
  stz player_facing
  stz player_dxlo
  lda #184
  sta player_yhi
  setaxy16
  stz player_frame_sub
  lda #48 << 8
  sta player_xlo

forever:

  jsl move_player

  ; Draw the player to a display list in main memory
  setaxy16
  stz oam_used
  jsl draw_player_sprite

  ; Mark remaining sprites as offscreen, then convert sprite size
  ; data from the convenient-to-manipulate format described by
  ; psycopathicteen to the packed format that the PPU actually uses.
  ldx oam_used
  jsl ppu_clear_oam
  jsl ppu_pack_oamhi

  ; Backgrounds and OAM can be modified only during vertical blanking.
  ; Wait for vertical blanking and copy prepared data to OAM.
  jsl ppu_vsync
  jsl ppu_copy_oam
  seta8
  lda #$0F
  sta PPUBRIGHT  ; turn on rendering

  ; wait for control reading to finish
  lda #$01
padwait:
  bit VBLSTATUS
  bne padwait
  stz BGSCROLLX
  stz BGSCROLLX

  jmp forever
.endproc





tools/wav2brr.py
#!/usr/bin/env python2
from __future__ import with_statement, division, print_function
from contextlib import closing
import array, wave, sys, os, optparse
try:
    import numpy as np
except ImportError:
    using_numpy = False
else:
    using_numpy = True

def naive_convolve(fircoeffs, wavdata):
    lside = len(fircoeffs) - 1
    rside = len(fircoeffs) - lside
    return [sum(a * b
                for a, b in zip(fircoeffs[max(lside - i, 0):],
                                wavdata[max(i - lside, 0): i + rside]))
            for i in xrange(len(wavdata) + len(fircoeffs) - 1)]

convolve = np.convolve if using_numpy else naive_convolve

def convolve_test():
    fircoeffs = [1/16, 4/16, 6/16, 4/16, 1/16]
    data = [0, 1, 0, 0, 0, 1, 2, 3, 4, 5]
    print(naive_convolve(fircoeffs, data))
    print(list(np.convolve(fircoeffs, data)))

def load_wave_as_mono_s16(filename):
    little = array.array('H', '\x01\x00')[0] == 1
    with closing(wave.open(filename, "rb")) as infp:
        bytedepth = infp.getsampwidth()
        if bytedepth not in (1, 2):
            raise ValueError("unsupported sampwidth")
        n_ch = infp.getnchannels()
        datatype = 'h' if bytedepth == 2 else 'B'
        freq = infp.getframerate()
        length = infp.getnframes()
        data = array.array('h', infp.readframes(length))
    if datatype == 'B':
        # Expand 8 to 16 bit
        data = array.array('h', ((c - 128) << 8 for c in data))
    elif not little:
        # 16-bit data is little-endian in the wave file; it needs to
        # be byteswapped for big-endian platforms
        data.byteswap()
    if n_ch > 1:
        # average all channels
        data = array.array('h', (int(round(sum(data[i:i + n_ch]) / n_ch))
                                 for i in xrange(0, len(data), n_ch)))
    return (freq, data)

def save_wave_as_mono_u16(filename, freq, data):
    data = array.array('h', (min(max(s, -32767), 32767) for s in data))
    little = array.array('H', '\x01\x00')[0] == 1
    if not little:
        data.byteswap()
    with closing(wave.open(filename, "wb")) as outfp:
        outfp.setnchannels(1)
        outfp.setsampwidth(2)
        outfp.setframerate(freq)
        outfp.writeframes(data.tostring())

def brr_preemphasize(wavdata):
    """Emphasize a sample's highs to compensate for S-DSP Gaussian interpolation."""
    preem = \
    [0.00417456, -0.017636, 0.03906008, -0.06200069, 0.07113042,
    -0.0255472, -0.26998287, 1.52160339, -0.26998287, -0.0255472,
    0.07113042, -0.06200069, 0.03906008, -0.017636, 0.00417456]
    return [int(round(s)) for s in convolve(preem, wavdata)[7:-7]]

def brr_deemphasize(wavdata):
    """Approximate the effect of Gaussian interpolation."""
    deem = \
    [372/2048, 1304/2048, 372/2048]
    return [int(round(s)) for s in convolve(deem, wavdata)[1:-1]]
    

brr_filters = [
    (0, 0),
    (0, 0.9375),
    (-0.9375, 1.90625),
    (-0.8125, 1.796875)
]

def enfilter(wav, coeffs, prevs=[], quant=None):
    """Calculates residuals from an IIR predictive filter.

wav: iterable of sample values
prevs: previous block of quantized and decoded samples
quant: None during test filtering; 2 to 4096 during encoding

"""
    prevs = list(prevs)[-len(coeffs):]
    if len(prevs) < len(coeffs):
        prevs = ([0] * len(coeffs) + prevs)[-len(coeffs):]
    out = []
    for c in wav:
        pred = sum(coeff * prev
                   for coeff, prev in zip(coeffs, prevs[-len(coeffs):]))
        rescaled = resid = (c - pred)
        if quant:
            resid = max(-8, min(7, int(round(resid / quant))))
            rescaled = resid * quant
        out.append(resid)
        prevs.append(rescaled + pred)
    return out, prevs[-len(coeffs):]

def defilter(wav, coeffs, prevs=[], quant=1):
    """Applies an IIR predictive filter to residuals.

wav: block of unpacked residuals
prevs: previous block of decoded samples
quant: number by which all residuals shall be multiplied

"""
    prevs = list(prevs)[-len(coeffs):]
    if len(prevs) < len(coeffs):
        prevs = ([0] * len(coeffs) + prevs)[-len(coeffs):]
    out = []
    for resid in wav:
        rescaled = resid * quant
        pred = sum(coeff * prev
                   for coeff, prev in zip(coeffs, prevs[-len(coeffs):]))
        c = (rescaled + int(round(pred)))
        out.append(c)
        prevs.append(c)
    return out

def encode_brr(wav, looped=False):
    prevs = []
    out = array.array('B')
    for t in xrange(0, len(wav), 16):
        piece = wav[t:t + 16]
        if prevs:
            # Calculate the peak residual for this piece with
            # each filter, then choose the smallest
            trials = [max(abs(resid)
                          for resid in enfilter(piece, coeffs, prevs)[0])
                      for coeffs in brr_filters]
            peak, filterid = min((r, i) for (i, r) in enumerate(trials))
        else:
            # first block always uses filter 0
            peak = max(abs(resid) for resid in piece)
            filterid = 0
        logquant = 0
        while logquant < 12 and peak > (15 << logquant):
            logquant += 1
        resids, prevs = enfilter(piece, brr_filters[filterid],
                                 prevs or [], 2 << logquant)
        resids.extend([0] * (16 - len(resids)))
        byte0 = (logquant << 4) | (filterid << 2)
        if t + 16 >= len(wav):
            byte0 = byte0 | (3 if looped else 1)
##        print("filter #%d, scale %d,\nresids %s\n%s"
##              % (filterid, 1 << logquant, repr(resids), repr(list(piece))))
        out.append(byte0)
        for i in xrange(0, len(resids), 2):
            hinibble = resids[i] & 0x0F
            lonibble = resids[i + 1] & 0x0F
            out.append((hinibble << 4) | lonibble)
    return out.tostring()

def decode_brr(brrdata):
    prevs = [0, 0]
    out = []
    for i in xrange(0, len(brrdata), 9):
        piece = array.array('B', brrdata[i:i + 9])
        logquant = piece[0] >> 4
        filterid = (piece[0] >> 2) & 0x03
        resids = [((b >> i & 0x0F) ^ 8) - 8
                  for b in piece[1:] for i in (4, 0)]
        decoded = defilter(resids, brr_filters[filterid],
                           prevs, 2 << logquant)
##        print("filter #%d, scale %d,\nresids %s\n%s"
##              % (filterid, 1 << logquant, repr(resids), repr(decoded)))
        out.extend(decoded)
        prevs = decoded
    return out

# Command line parsing and help #####################################
#
# Easy is hard.  It takes a lot of code and a lot of text to make a
# program self-documenting and resilient to bad input.

usageText = "usage: %prog [options] [-i] INFILE [-o] OUTFILE"
versionText = """wav2brr 0.04

Copyright 2014 Damian Yerrick
Copying and distribution of this file, with or without
modification, are permitted in any medium without royalty provided
the copyright notice and this notice are preserved in all source
code copies. This file is offered as-is, without any warranty.
"""
descriptionText = """
Audio converter for Super NES S-DSP.
"""

def parse_argv(argv):
    parser = optparse.OptionParser(usage=usageText, version=versionText,
                                   description=descriptionText)
    parser.add_option("-i", dest="infilename",
                      help="read from INFILE",
                      metavar="INFILE")
    parser.add_option("-o", dest="outfilename",
                      help="write output to OUTFILE",
                      metavar="OUTFILE")
    parser.add_option("-d", "--decompress",
                      action="store_true", dest="decompress", default=False,
                      help="decompress BRR to wave (default: compress wave to BRR)")
    parser.add_option("-r", "--rate", dest="rate",
                      metavar="RATE", type="int", default=8372,
                      help="set the wave sample rate in Hz (default: 8372)")

    (options, pos) = parser.parse_args(argv[1:])

    if not 10 <= options.rate <= 128000:
        parser.error("output sample rate must be 10 to 128000 Hz")

    # Fill unfilled roles with positional arguments
    pos = iter(pos)
    try:
        options.infilename = options.infilename or pos.next()
    except StopIteration:
        parser.error("no input file; try %s --help"
                     % os.path.basename(sys.argv[0]))
    try:
        options.outfilename = options.outfilename or pos.next()
    except StopIteration:
        parser.error("no output file")

    # make sure no trailing arguments
    try:
        pos.next()
        parser.error("too many filenames")
    except StopIteration:
        pass
    return options

def main(argv=None):
    argv = argv or sys.argv
    opts = parse_argv(argv)
    if opts.decompress:
        with open(opts.infilename, 'rb') as infp:
            brrdata = infp.read()
        out = brr_deemphasize(decode_brr(brrdata))
        save_wave_as_mono_u16(opts.outfilename, opts.rate, out)
    else:
        freq, wave = load_wave_as_mono_s16(opts.infilename)
        brrdata = encode_brr(brr_preemphasize(wave))
        with open(opts.outfilename, 'wb') as outfp:
            outfp.write(brrdata)

if __name__=='__main__':
    main()
##    convolve_test()


tools/mktables.py
#!/usr/bin/env python
from __future__ import print_function, with_statement, division
import math
try:
    xrange
except NameError:
    xrange = range

def make_bytetable(it):
    x = 0
    out = []
    for el in it:
        out.append(',' if x else '  .byte ')
        out.append("%3d" % el)
        x += 1
        if x >= 16:
            x = 0
            out.append('\n')
    if x:
        out.append('\n')
    return ''.join(out)

# Playback frequency on the Super NES DSP is expressed in
# 125/16 = 7.8125 Hz units and can go no higher than 128 kHz.
# Our tuning assumes that each sample's period is 32 units,
# or middle C at (261+5/8) * 32 = 8372 Hz.

lowest_freq = 55 * 32 / 7.8125
num_notes = 64
freqs = [int(round(lowest_freq * math.pow(2, i / 12.0)))
          for i in xrange(num_notes)]

tmpl = """; Lookup tables for Super NES sound engine
; generated with mktables.py

.segment "SPCIMAGE"
noteFreqsLo:
%s
noteFreqsHi:
%s
"""
tables = (
    make_bytetable(f & 0xFF for f in freqs),
    make_bytetable(f >> 8 for f in freqs)
)
    
print(tmpl % tables)


src/snesheader.s
;
; Definition of the internal header and vectors at $00FFC0-$00FFFF
; in the Super NES address space.
;

.include "snes.inc"
;.include "global.inc"
.smart
.import resetstub, main, nmi_handler, irq_handler
.export map_mode:far

; LoROM is mapped to $808000-$FFFFFF in 32K blocks,
; skipping a15 and a23. Most is mirrored down to $008000.
MAPPER_LOROM = $20    ; a15 skipped
; HiROM is mapped to $C00000-$FFFFFF linearly.
; It is mirrored down to $400000, and the second half of
; each 64K bank is mirrored to $008000 and $808000.
MAPPER_HIROM = $21    ; C00000-FFFFFF skipping a22, a23
; ExHiROM is mapped to $C00000-$FFFFFF followed by
; $400000-$7FFFFF.  There are two inaccessible 32K holes
; near the end, and two 32K blocks that are accessible only
; through their mirrors to $3E8000 and $3F8000.
MAPPER_EXHIROM = $25  ; skipping a22, inverting a23

; If ROMSPEED_120NS is turned on, the game will be pressed
; on more expensive "fast ROM" that allows the CPU to run
; at its full 3.6 MHz when accessing ROM at $808000 and up.
ROMSPEED_200NS = $00
ROMSPEED_120NS = $10

; ROM and backup RAM sizes are expressed as
; ceil(log2(size in bytes)) - 10
MEMSIZE_NONE  = $00
MEMSIZE_2KB   = $01
MEMSIZE_4KB   = $02
MEMSIZE_8KB   = $03
MEMSIZE_16KB  = $04
MEMSIZE_32KB  = $05
MEMSIZE_64KB  = $06
MEMSIZE_128KB = $07  ; values from here up are for SRAM
MEMSIZE_256KB = $08  ; values from here down are for ROM
MEMSIZE_512KB = $09  ; Super Mario World
MEMSIZE_1MB   = $0A  ; Mario Paint
MEMSIZE_2MB   = $0B  ; Mega Man X, Super Mario All-Stars, original SF2
MEMSIZE_4MB   = $0C  ; Turbo/Super SF2, all Donkey Kong Country
MEMSIZE_8MB   = $0D  ; ExHiROM only: Tales of Phantasia, SF Alpha 2

REGION_JAPAN = $00
REGION_AMERICA = $01
REGION_PAL = $02

.segment "SNESHEADER"
romname:
  ; The ROM name must be no longer than 21 characters.
  ; Longest possible is "PINO'S LOROM TEMPLATE"
  .byte "PINO'S LOROM TEMPLATE"
  .assert * - romname <= 21, error, "ROM name too long"
  .if * - romname < 21
    .res romname + 21 - *, $20  ; space padding
  .endif
map_mode:
  .byte MAPPER_LOROM|ROMSPEED_200NS
  .byte $00   ; 00: no extra RAM; 02: RAM with battery
  .byte MEMSIZE_256KB  ; ROM size (08-0C typical)
  .byte MEMSIZE_NONE   ; backup RAM size (01,03,05 typical; Dezaemon has 07)
  .byte REGION_AMERICA
  .byte $33   ; publisher id, or $33 for see 16 bytes before header
  .byte $00   ; ROM revision number
  .word $0000 ; sum of all bytes will be poked here after linking
  .word $0000 ; $FFFF minus above sum will also be poked here
  .res 4  ; unused vectors
  ; clcxce mode vectors
  ; reset unused because reset switches to 6502 mode
  .addr cop_handler, brk_handler, abort_handler
  .addr nmistub, $FFFF, irq_handler
  .res 4  ; more unused vectors
  ; 6502 mode vectors
  ; brk unused because 6502 mode uses irq handler and pushes the
  ; X flag clear for /IRQ or set for BRK
  .addr ecop_handler, $FFFF, eabort_handler
  .addr enmi_handler, resetstub, eirq_handler
  
.segment "CODE"

; Jumping out of bank $00 is especially important if you're using
; ROMSPEED_120NS.
nmistub:
  jml nmi_handler

irqstub:
  jml irq_handler

; Unused exception handlers
cop_handler:
brk_handler:
abort_handler:
ecop_handler:
eabort_handler:
enmi_handler:
eirq_handler:
  rti

